gufe/storage/externalresource.py

```diff
@@ -0,0 +1,301 @@
+        """
+        return _ForceContext(self._load_stream(location))
+
+    def delete(self, location):
+        """
+        Delete an existing data chunk from the backend.
+
+        Subclasses should implement ``_delete``.
+
+        Parameters
+        ----------
+        location : str
+            label for the data to delete
+
+        Raises
+        ------
+        MissingExternalResourceError
+            If the resource to be deleted does not exist
+        """
+        return self._delete(location)
+
+    def store_bytes(self, location, byte_data):
+        """
+        Store given data in the backend.
+
+        Subclasses should implement ``_store``.
+
+        Parameters
+        ----------
+        location : str
+            label associated with the data to store
+        byte_data : bytes
+            bytes to store
+        """
+        return self._store_bytes(location, byte_data)
+
+    def store_path(self, location, path: pathlib.Path):
+        self._store_path(location, path)
+
+    def exists(self, location) -> bool:
+        """
+        Check whether a given label has already been used.
+
+        Subclasses should implement ``_exists``.
+
+        Parameters
+        ----------
+        location : str
+            the label to check for
+
+        Return
+        ------
+        bool
+            True if this label has associated data in the backend, False if
+            not
+        """
+        return self._exists(location)
+
+    @abc.abstractmethod
+    def iter_contents(self, prefix=""):
+        """Iterate over the labels in this storage.
+
+        Parameters
+        ----------
+        prefix : str
+            Only iterate over paths that start with the given prefix.
+
+        Returns
+        -------
+        Iterator[str] :
+            Contents of this storage, which may include items without
+            metadata.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def _store_bytes(self, location, byte_data):
+        """
+        For implementers: This should be blocking, even if the storage
+        backend allows asynchronous storage.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def _store_path(self, location, path):
+        """
+        For implementers: This should be blocking, even if the storage
+        backend allows asynchronous storage.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def _exists(self, location) -> bool:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def _delete(self, location):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def _get_filename(self, location):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def _load_stream(self, location):
+        raise NotImplementedError()
+
+
+# TODO: this should use pydantic to check init inputs
+class FileStorage(ExternalStorage):
```

Comment:
Might be better to split this out into another file, same with MemoryStorage below